File: 139-word-break/139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        cache = []
        def dp(rem: str):
            for word in wordDict:
                if rem in cache:
                    return False
                if rem == word:
                    return True
                if rem[0:len(word)] == word:
                    if dp(rem[len(word):]) is True:
                        return True
            cache.append(rem)
            return False
        return dp(s)
        

File: 139-word-break/139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        auto len_s = s.size();
        vector<bool> dp (len_s+1, false);
        dp.back() = true;
        for (int i = len_s-1; i>=0; i--) {
            for (string word : wordDict) {
                if (!dp[i] && i+word.size() <= len_s && s.substr(i, word.size()) == word) {
                    dp[i] = dp[i + word.size()];
                }
            }
        }
        return dp[0];
    }
};